### PR TITLE
Fix 154: adding a finalizer to the tunnel's associated secret

### PR DESCRIPTION
**Pull request by cyclingwithelephants**
fixes https://github.com/adyanth/cloudflare-operator/issues/154

I just wanted to confirm that `r.GetTunnel().GetNamespace()` for the ClusterTunnel resource is expected to return the cloudflare-operator's namespace? I've tested this working, just wanted to make sure this was built on expected behaviour and wasn't a fluke

### CODE REVIEW

**Comment by cyclingwithelephants** [comment]
Looks like coverage reports are not available from forks

File: internal/k8s/secret.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		return nil, errors.New("logger cannot be nil")
+	}
+	return &SecretClient{
+		k8sClient: k8sClient,
+		log:       log,
+	}, nil
+}
+
+// EnsureFinalizer idempotently adds the specified finalizer to the specified secret
+func (s *SecretClient) EnsureFinalizer(ctx context.Context, secretName, secretNamespace, finalizer string) error {
```
Comment (adyanth) [inline internal/k8s/secret.go:32]: Could you accept a namespacedName instead of two args? Same elsewhere

File: internal/k8s/secret.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	}
+
+	s.log.WithValues("finalizer", finalizer).Info("creating finalizer")
+	secret.Finalizers = append(secret.Finalizers, finalizer)
+	return s.k8sClient.Update(ctx, &secret)
+}
+
+// RemoveFinalizer removes the first instance of the finalizer from the given secret
+func (s *SecretClient) RemoveFinalizer(ctx context.Context, secretName, secretNamespace, finalizer string) error {
```
Comment (adyanth) [inline internal/k8s/secret.go:55]: Looks like these are generic methods that can apply to any object. Instead of making a secrets client, could we have this function accept a resource type and query for that and add/remove finalizer? Would help reuse this for configmaps too.

File: internal/k8s/secret.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// removeString returns a copy of list with the first (only) occurrence
+// of target removed. If target is not present, the original slice is
+// returned unchanged.
+func removeString(list []string, target string) []string {
+	for i, v := range list {
+		if v == target {
+			// splice out the element at index i
+			return append(list[:i], list[i+1:]...)
```
Comment (adyanth) [inline internal/k8s/secret.go:80]: could you add some tests for this? Does this work for corner cases? Like removing last element?

**Review by adyanth — CHANGES_REQUESTED** [review]
Thanks for picking this up!

**Comment by adyanth** [comment]
To answer your question, yes the namespace for the cluster tunnel is the operator namespace.

File: internal/k8s/secret.go (hunk 1 above)
Comment (cyclingwithelephants) [reply internal/k8s/secret.go]: I'm accepting a client.ObjectKey as otherwise it would require a conversion in the function

File: internal/k8s/secret.go (hunk 2 above)
Comment (cyclingwithelephants) [reply internal/k8s/secret.go]: done :)

File: internal/k8s/secret.go (hunk 3 above)
Comment (cyclingwithelephants) [reply internal/k8s/secret.go]: removed this in favour of inbuilt library functions I found

**Comment by cyclingwithelephants** [comment]
Haven't tested yet, but please let me know if you're otherwise happy with the changes

File: internal/k8s/object.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	base, err := s.deepCopyObject(obj)
+	if err != nil {
+		return err
+	}
+
+	s.log.
+		WithValues("finalizer", finalizer).
+		WithValues("object", fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName())).
+		Info("creating finalizer")
+	if err := s.k8sClient.Patch(ctx, obj, client.MergeFrom(base)); err != nil {
```
Comment (adyanth) [inline internal/k8s/object.go:57]: Any way to share this with the existing Merge/Apply in `kubectl_apply.go`? Ok to resolve if not too easy.

File: internal/k8s/object.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+func (*ObjectClient) deepCopyObject(obj client.Object) (client.Object, error) {
+	objDeepCopy := obj.DeepCopyObject()
+	if objDeepCopy == nil {
+		return nil, errors.New("received nil object from DeepCopyObject")
+	}
+	base, ok := objDeepCopy.(client.Object)
+	if !ok {
+		return nil, errors.New("failed to convert object to client.Object")
+	}
```
Comment (adyanth) [inline internal/k8s/object.go:103]: Is there anything that would cause a deep copied `client.Object` to not return a `client.Object`? Or do you want to switch to `runtime.Object` instead if we do not use `meta.Object` anywhere?

The error checks seem probably redundant unless I'm missing something. It would be nice if a deep copy never failed (returning nil for a deep copy of nil is valid). 

I would think instead of this function, you could call obj.DeepCopyObject() where needed and nil checks won't be necessary since at the call sites, you've already validated that the obj is not nil.

**Review by adyanth — COMMENTED** [review]
Couple of suggestions, looks good otherwise!

File: internal/k8s/object.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+// EnsureFinalizer adds `finalizer` to obj exactly once.
+//   - It NO-OPs if the finalizer is already there.
+//   - It uses a strategic-merge Patch so you don’t overwrite concurrent changes.
+//   - obj must be a pointer that already contains the latest copy from the API server.
+func (s *ObjectClient) EnsureFinalizer(ctx context.Context, key client.ObjectKey, finalizer string) error {
+	var obj client.Object
+	err := s.k8sClient.Get(ctx, key, obj)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (cyclingwithelephants) [reply internal/k8s/object.go:40]: yeah I haven't tested it yet, I'm not very sure this will work

File: internal/k8s/object.go (hunk 5 above)
Comment (cyclingwithelephants) [reply internal/k8s/object.go:103]: IIRC this was a lint error, since doesn't know it can't be nil. I figured it was safer (albeit definitely not cleaner) to add the nil check than overrule the linter but we can do that too

File: internal/k8s/object.go (hunk 4 above)
Comment (cyclingwithelephants) [reply internal/k8s/object.go:57]: yeah I'll have a look into this

File: internal/k8s/object.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (adyanth) [reply internal/k8s/object.go:40]: I was thinking the caller supplies a client.Object with a concrete type. Like `addFinalizer(..., &corev1.Secret{})`

File: internal/k8s/object.go (hunk 5 above)
Comment (adyanth) [reply internal/k8s/object.go:103]: Does the linter complain if you do this at the call site? I **think** it should be smart enough to know if there is no err, the obj cant be nil?

File: internal/k8s/object.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cyclingwithelephants) [reply internal/k8s/object.go:40]: Oh I see, yeah that's definitely sensible. Let me do some testing and see if it turns out to be magic (I'm sure it won't)

File: internal/k8s/object.go (hunk 5 above)
Comment (cyclingwithelephants) [reply internal/k8s/object.go:103]: I'll have a play :)

Comment (cyclingwithelephants) [reply internal/k8s/object.go]: Decided to just do the type assertion inline, we know better than the compiler here

File: internal/k8s/object.go (hunk 4 above)
Comment (cyclingwithelephants) [reply internal/k8s/object.go]: okay I had a look at this, and my gut feeling is that not having this depending on the genericReconciler is a good thing, it's pretty self contained as it stands and I think we can do work to break down the genericReconciler interface into being multiple leaner interfaces. What do you think?

**Comment by cyclingwithelephants** [comment]
this is functionally working, but I've noticed that the controller is producing errors on deletion. This wasn't happening when using the concrete implementation for the secret. What's strange is that this `Reconciler error` (the last error in the logs) loops maybe 15 times and it looks like it's happening after the tunnel is actually deleted from etcd. Any idea what's going on here?

```
2025-05-08T23:35:54+01:00	INFO	starting deletion cycle	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "2a666e68-bfb5-43e7-bdb2-d4bcc8da318d"}
2025-05-08T23:35:54+01:00	INFO	Scaling down cloudflared	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "2a666e68-bfb5-43e7-bdb2-d4bcc8da318d"}
2025-05-08T23:35:54+01:00	DEBUG	events	Starting Tunnel Deletion	{"type": "Normal", "object": {"kind":"Tunnel","namespace":"default","name":"adamrummer-com","uid":"c20c36ca-7fa3-408f-9e60-b679c7085b02","apiVersion":"networking.cfargotunnel.com/v1alpha2","resourceVersion":"2001"}, "reason": "Deleting"}
2025-05-08T23:35:54+01:00	DEBUG	events	Scaling down cloudflared	{"type": "Normal", "object": {"kind":"Tunnel","namespace":"default","name":"adamrummer-com","uid":"c20c36ca-7fa3-408f-9e60-b679c7085b02","apiVersion":"networking.cfargotunnel.com/v1alpha2","resourceVersion":"2001"}, "reason": "Scaling"}
2025-05-08T23:35:54+01:00	INFO	Scaling down successful	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "2a666e68-bfb5-43e7-bdb2-d4bcc8da318d"}
2025-05-08T23:35:54+01:00	DEBUG	events	Scaling down cloudflared successful	{"type": "Normal", "object": {"kind":"Tunnel","namespace":"default","name":"adamrummer-com","uid":"c20c36ca-7fa3-408f-9e60-b679c7085b02","apiVersion":"networking.cfargotunnel.com/v1alpha2","resourceVersion":"2001"}, "reason": "Scaled"}
2025-05-08T23:35:54+01:00	INFO	starting deletion cycle	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "e7e01bed-c04d-48d9-bde0-53744e6ce628"}
2025-05-08T23:35:54+01:00	INFO	In validation	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "e7e01bed-c04d-48d9-bde0-53744e6ce628"}
2025-05-08T23:35:54+01:00	DEBUG	events	Starting Tunnel Deletion	{"type": "Normal", "object": {"kind":"Tunnel","namespace":"default","name":"adamrummer-com","uid":"c20c36ca-7fa3-408f-9e60-b679c7085b02","apiVersion":"networking.cfargotunnel.com/v1alpha2","resourceVersion":"2001"}, "reason": "Deleting"}
2025-05-08T23:35:54+01:00	INFO	no TXT records returned for fqdn	{"controller": "tunnelbinding", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "TunnelBinding", "TunnelBinding": {"name":"ingress-nginx","namespace":"default"}, "namespace": "default", "name": "ingress-nginx", "reconcileID": "72ca0583-cfc1-4a8c-aec0-c39310e4d300", "fqdn": "*.adamrummer.com"}
2025-05-08T23:35:55+01:00	INFO	no records returned for fqdn	{"controller": "tunnelbinding", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "TunnelBinding", "TunnelBinding": {"name":"ingress-nginx","namespace":"default"}, "namespace": "default", "name": "ingress-nginx", "reconcileID": "72ca0583-cfc1-4a8c-aec0-c39310e4d300", "fqdn": "*.adamrummer.com"}
2025-05-08T23:35:55+01:00	ERROR	Error fetching DNS record	{"controller": "tunnelbinding", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "TunnelBinding", "TunnelBinding": {"name":"ingress-nginx","namespace":"default"}, "namespace": "default", "name": "ingress-nginx", "reconcileID": "72ca0583-cfc1-4a8c-aec0-c39310e4d300", "Hostname": "*.adamrummer.com", "error": "no records returned"}
github.com/adyanth/cloudflare-operator/internal/controller.(*TunnelBindingReconciler).deleteDNSLogic
	/Users/adam/code/github.com/adyanth/cloudflare-operator/internal/controller/tunnelbinding_controller.go:370
github.com/adyanth/cloudflare-operator/internal/controller.(*TunnelBindingReconciler).deletionLogic
	/Users/adam/code/github.com/adyanth/cloudflare-operator/internal/controller/tunnelbinding_controller.go:230
github.com/adyanth/cloudflare-operator/internal/controller.(*TunnelBindingReconciler).Reconcile
	/Users/adam/code/github.com/adyanth/cloudflare-operator/internal/controller/tunnelbinding_controller.go:174
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:119
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:334
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:294
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:255
2025-05-08T23:35:55+01:00	DEBUG	events	Error fetching DNS record	{"type": "Warning", "object": {"kind":"TunnelBinding","namespace":"default","name":"ingress-nginx","uid":"512c1f11-02dc-427c-a94e-20d2384bc8ad","apiVersion":"networking.cfargotunnel.com/v1alpha1","resourceVersion":"2002"}, "reason": "FailedDeletingDns"}
2025-05-08T23:35:55+01:00	DEBUG	events	Finalizer removed	{"type": "Normal", "object": {"kind":"TunnelBinding","namespace":"default","name":"ingress-nginx","uid":"512c1f11-02dc-427c-a94e-20d2384bc8ad","apiVersion":"networking.cfargotunnel.com/v1alpha1","resourceVersion":"2002"}, "reason": "FinalizerUnset"}
2025-05-08T23:35:55+01:00	INFO	TunnelBinding deleted, nothing to do	{"controller": "tunnelbinding", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "TunnelBinding", "TunnelBinding": {"name":"ingress-nginx","namespace":"default"}, "namespace": "default", "name": "ingress-nginx", "reconcileID": "ffaf97d1-676c-4680-8371-8fe4fc58d0bd"}
2025-05-08T23:35:56+01:00	INFO	Tunnel deleted	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "e7e01bed-c04d-48d9-bde0-53744e6ce628", "tunnelID": "8447f5e8-161d-407e-a721-5eda624d3600"}
2025-05-08T23:35:56+01:00	DEBUG	events	Tunnel deletion successful	{"type": "Normal", "object": {"kind":"Tunnel","namespace":"default","name":"adamrummer-com","uid":"c20c36ca-7fa3-408f-9e60-b679c7085b02","apiVersion":"networking.cfargotunnel.com/v1alpha2","resourceVersion":"2001"}, "reason": "Deleted"}
2025-05-08T23:35:56+01:00	DEBUG	events	Tunnel Finalizer removed	{"type": "Normal", "object": {"kind":"Tunnel","namespace":"default","name":"adamrummer-com","uid":"c20c36ca-7fa3-408f-9e60-b679c7085b02","apiVersion":"networking.cfargotunnel.com/v1alpha2","resourceVersion":"2001"}, "reason": "FinalizerUnset"}
2025-05-08T23:35:56+01:00	ERROR	Reconciler error	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "e7e01bed-c04d-48d9-bde0-53744e6ce628", "error": "Operation cannot be fulfilled on tunnels.networking.cfargotunnel.com \"adamrummer-com\": StorageError: invalid object, Code: 4, Key: /registry/networking.cfargotunnel.com/tunnels/default/adamrummer-com, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: c20c36ca-7fa3-408f-9e60-b679c7085b02, UID in object meta: "}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:347
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:294
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:255
2025-05-08T23:35:56+01:00	INFO	removing finalizer	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "3ab2e802-ae85-479d-ace1-43a7a949458d", "finalizer": "cfargotunnel.com/finalizer", "object": "default/cloudflare-secrets"}
2025-05-08T23:35:56+01:00	ERROR	Reconciler error	{"controller": "tunnel", "controllerGroup": "networking.cfargotunnel.com", "controllerKind": "Tunnel", "Tunnel": {"name":"adamrummer-com","namespace":"default"}, "namespace": "default", "name": "adamrummer-com", "reconcileID": "7e29d1a4-44ed-4ed7-9c05-6def336a799f", "error": "Secret \"cloudflare-secrets\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:347
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:294
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2
	/Users/adam/.cache/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.20.4/pkg/internal/controller/controller.go:255
```

File: internal/controller/clustertunnel_controller.go [hunk 7]
```diff
@@ -145,9 +147,25 @@ func (r *ClusterTunnelReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	if err := r.Get(ctx, req.NamespacedName, tunnel); err != nil {
 		if apierrors.IsNotFound(err) {
 			// Tunnel object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			r.log.Info("Tunnel deleted, nothing to do")
+			// Owned objects are automatically garbage collected.
+			objectClient, err := k8s.NewObjectClient(r.Client, &r.log)
+			if err != nil {
+				return ctrl.Result{}, err
```
Comment (adyanth) [inline internal/controller/clustertunnel_controller.go:153]: This logic is in the wrong place (ref: the error you saw). When this condition is hit, the tunnel is already removed from etcd. 

What you need is, instead of the logic in both here and `tunnel_controller.go`, move this to https://github.com/adyanth/cloudflare-operator/blob/d1570e5b098d688258e1b60f6fac44e25357d2d8/internal/controller/generic_tunnel_reconciler.go#L157-L162

Comment (cyclingwithelephants) [reply internal/controller/clustertunnel_controller.go:153]: Oh, derp. Thanks will fix this

Comment (adyanth) [reply internal/controller/clustertunnel_controller.go:153]: Did you get around to this? Would love to merge this in

Comment (adyanth) [reply internal/controller/clustertunnel_controller.go:153]: All good. Life comes first!

Comment (cyclingwithelephants) [reply internal/controller/clustertunnel_controller.go]: Just wanted to give you another heads up, might be another week
